### PR TITLE
Adds basic local version of tracked state system

### PR DESCRIPTION
**Pull request by Spoffy**

### CODE REVIEW

File: paradigm/global/functions/tracked_state/fn_trState_whenEqual.sqf [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+private _stateEqualsHandlers = _trStateData get "equalsHandlers";
+private _variableStateHandlers = _stateEqualsHandlers getOrDefault [_stateVariable, createHashMap, true];
+private _matchingValueHandlers = _variableStateHandlers getOrDefault [_valueHash, [], true];
+
+private _handlerId = ["equalsHandlers", _stateVariable, _valueHash, _callback];
+private _handler = [_handlerId, _callback];
+_matchingValueHandlers pushBack _handler;
+
+private _currentValue = _trStateData get "state" get _stateVariable;
+if (!isNil "_currentValue" && { hashValue _currentValue isEqualTo hashValue _valueToMatch }) then {
```
Comment (Spoffy) [inline paradigm/global/functions/tracked_state/fn_trState_whenEqual.sqf]: Can use _valueHash on this line

File: paradigm/global/functions/tracked_state/fn_trState_callHandlers.sqf [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    Author: Savage Game Design
+    Date: 2023-02-09
+    Last Update: 2023-02-10
+    Public: No
+
+    Description:
+        Calls the given handlers with the given parameters.
+
+    Parameter(s):
+        _handlers - Handlers to call [ARRAY]
```
Comment (gnalvesteffer) [inline paradigm/global/functions/tracked_state/fn_trState_callHandlers.sqf]: I think it'd be helpful to describe the schema for the handlers, e.g. `[ARRAY of [HANDLER ID, HANDLER CODE]]`. (at least I think that's what the schema is from what I've read in other parts of the PR)

File: paradigm/global/functions/tracked_state/fn_trState_callHandlers.sqf [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        _stateVariable - State variable that's being observed [ANY HASHMAP KEY]
+        _currentValue - Current value of the state variable [ANY HASHABLE]
+        _oldValue - Old value of the state variable, may be nil [ANY HASHABLE]
+
+    Returns:
+        Nothing
+
+    Example(s):
+        [
+            _equalsHandlers,
```
Comment (gnalvesteffer) [inline paradigm/global/functions/tracked_state/fn_trState_callHandlers.sqf]: To my previous comment about describing the handler schema, it'd be nice to elaborate this example as well, showing off said schema.

File: paradigm/global/functions/tracked_state/fn_trState_data.sqf [hunk 4]
```diff
@@ -0,0 +1,31 @@
+/*
+    File: fn_trstate_data.sqf
```
Comment (gnalvesteffer) [inline paradigm/global/functions/tracked_state/fn_trState_data.sqf]: Not sure if this is the convention for this codebase, but I'd personally recommend renaming this to `fn_trstate_getdata.sqf`, as it makes the operation more clear IMO.

File: paradigm/global/functions/tracked_state/fn_trState_setLocal.sqf [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    Example(s):
+        ["Local_GamemodeStage", "Starting"] call para_g_fnc_trstate_setLocal;
+ */
+
+params ["_stateVariable", "_newValue"];
+
+private _trStateData = [] call para_g_fnc_trState_data;
+
+private _state = _trstateData get "state";
```
Comment (gnalvesteffer) [inline paradigm/global/functions/tracked_state/fn_trState_setLocal.sqf]:
```suggestion
private _state = _trStateData get "state";
```

I don't remember if SQF variables are case sensitive, but this has a different casing than _trStateData.

Comment (gnalvesteffer) [reply paradigm/global/functions/tracked_state/fn_trState_setLocal.sqf]: Ah, looks like they're case-insensitive https://community.bistudio.com/wiki/Identifier:
image

File: paradigm/global/functions/tracked_state/fn_trState_whenEqual.sqf [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    Parameter(s):
+        _stateVariable - Variable to watch [ANY HASHMAP KEY]
+        _valueToMatch - Value to check for equality against [ANY HASHABLE]
+        _callback - Callback, with parameters [ARRAY]
+
+    Returns:
+        ID which can be used to remove the handler [ANY]
+
+    Example(s):
+        ["Local_GamemodeStage", "Starting", [[32], { DO THINGS }]] call para_g_fnc_trstate_whenChanged
```
Comment (gnalvesteffer) [inline paradigm/global/functions/tracked_state/fn_trState_whenEqual.sqf]: For this example, I think it'd be helpful to show usage of the incoming args that get passed in from `callHandlers` https://github.com/Savage-Game-Design/Tour-of-Duty/pull/16/files#diff-a235671d3cc3b8f9eda1d5c86be0b7fd19e4e4acb7fa3f8579d35e55a0841befR33

Comment (gnalvesteffer) [inline paradigm/global/functions/tracked_state/fn_trState_whenEqual.sqf]:
```suggestion
        ["Local_GamemodeStage", "Starting", [[32], { DO THINGS }]] call para_g_fnc_trstate_whenEqual
```

Function name doesn't match the script name `whenEqual` (I assume you renamed this a few times during development :P)

**Review by gnalvesteffer — APPROVED** [review]
Just a few comments primarily regarding ways to improve examples, but everything else looks good to me.

File: paradigm/global/functions/tracked_state/fn_trState_setLocal.sqf [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+// Shouldn't matter if this is nil, as long as it's only used in arrays
+private _oldValue = _state get _stateVariable;
+
+_state set [_stateVariable, _newValue];
+
```
Comment (veteran29) [inline paradigm/global/functions/tracked_state/fn_trState_setLocal.sqf]: Shouldn't this check if `_oldValue` is different than `_newValue`, what if we set same state multiple times?

File: paradigm/global/functions/tracked_state/fn_trState_setLocal.sqf (hunk 5 above)
Comment (Spoffy) [reply paradigm/global/functions/tracked_state/fn_trState_setLocal.sqf]: Yeah, they aren't, but it's good practice to keep it consistent anyway, so thanks for pointing it out :)

File: paradigm/global/functions/tracked_state/fn_trState_whenEqual.sqf (hunk 6 above)
Comment (Spoffy) [reply paradigm/global/functions/tracked_state/fn_trState_whenEqual.sqf]: Copy and pasting is a curse 😆

File: paradigm/global/functions/tracked_state/fn_trState_setLocal.sqf (hunk 7 above)
Comment (Spoffy) [reply paradigm/global/functions/tracked_state/fn_trState_setLocal.sqf]: That's a good thought, and is definitely how this *should* logically function - I'll do that.

File: paradigm/global/functions/tracked_state/fn_trackedState_setLocal.sqf [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+params ["_stateVariable", "_newValue"];
+
+private _trStateData = [] call para_g_fnc_trackedState_getData;
+
+private _state = _trStateData get "state";
+// Shouldn't matter if this is nil, as long as it's only used in arrays
+private _oldValue = _state get _stateVariable;
+
+// Use hash value comparison, as that's how equality is checked in whenEqual
+if (hashValue _oldValue isEqualTo hashValue _newValue) exitWith {};
```
Comment (veteran29) [inline paradigm/global/functions/tracked_state/fn_trackedState_setLocal.sqf:31]: This does nothing if _oldValue is a `nil` but it should not matter.

**Pull request merged**